src/diffusers/textual_inversion_utils.py

```diff
@@ -0,0 +1,101 @@
+      encoder's embedding matrix
+    """
+    textual_inversion_tokens = []
+
+    def load_textual_inversion_embeddings(self, embeddings):
+        r"""
+        Loads textual inversion embeddings.
+
+        Receives a dictionary with the following keys:
+        - `token`: name of the token to be added to the tokenizers' vocabulary
+        - `embedding`: path to the embedding of the token to be added to the text encoder's embedding matrix
+
+        Alternatively, it can receive a list of pathes to embedding dictionaries, where the keys are the tokens and the
+        values are the embeddings. In that case, it will iterate through the list and add the tokens and embeddings to
+        the tokenizer's vocabulary and the text encoder's embedding matrix.
+
+        Iters through the dictionary and adds the token to the tokenizer's vocabulary and the embedding to the text
+        encoder's embedding matrix.
+        """
+
+        if isinstance(embeddings, dict):
+            for token, embedding_path in embeddings.items():
+                # check if token in tokenizer vocab
+                # if yes, raise exception
+                if token in self.tokenizer.get_vocab():
+                    raise ValueError(
+                        f"Token {token} already in tokenizer vocabulary. Please choose a different token name."
+                    )
+
+                embedding_dict = torch.load(embedding_path)
+                embedding = self.extract_embedding_from_dict(embedding_dict)
+
+                self.add_textual_inversion_embedding(token, embedding)
+
+        elif isinstance(embeddings, list):
+            for embedding_path in embeddings:
+                embedding_dict = torch.load(embedding_path)
+                token = self.extract_token_from_dict(embedding_dict)
+                embedding = self.extract_embedding_from_dict(embedding_dict)
+
+                # check if token in tokenizer vocab
+                # if yes, raise exception
+                if token in self.tokenizer.get_vocab():
+                    raise ValueError(
+                        f"Token {token} already in tokenizer vocabulary. Please choose a different token name."
+                    )
+                self.add_textual_inversion_embedding(token, embedding)
+
+    def extract_embedding_from_dict(self, embedding_dict):
+        r"""
+        Extracts the embedding from the embedding dictionary.
+        """
+        # auto1111 embedding case
+        if "string_to_param" in embedding_dict:
+            embedding_dict = embedding_dict["string_to_param"]
+            embedding = embedding_dict["*"]
+            return embedding
+
+        return list(embedding_dict.values())[0]
+
+    def extract_token_from_dict(self, embedding_dict):
+        r"""
+        Extracts the token from the embedding dictionary.
+        """
+        # auto1111 embedding case
+        if "string_to_param" in embedding_dict:
```

Comment:
Interesting that this is the auto1111 format.

If I understand correctly a "string: string" mapping is saved in the embedding dict. This wouldn't work with `safetensors` which is a bit unfortunate :-/

Reply:
I can implement the safetensors case as well.